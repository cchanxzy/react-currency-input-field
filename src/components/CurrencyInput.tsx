import React, {
  FC,
  useState,
  useEffect,
  useRef,
  forwardRef,
  useMemo,
  useImperativeHandle,
} from 'react';
import { CurrencyInputProps, CurrencyInputOnChangeValues } from './CurrencyInputProps';
import {
  isNumber,
  cleanValue,
  fixedDecimalValue,
  formatValue,
  getLocaleConfig,
  padTrimValue,
  CleanValueOptions,
  getSuffix,
  FormatValueOptions,
  repositionCursor,
} from './utils';

export const CurrencyInput: FC<CurrencyInputProps> = forwardRef<
  HTMLInputElement,
  CurrencyInputProps
>(
  (
    {
      allowDecimals = true,
      allowNegativeValue = true,
      id,
      name,
      className,
      customInput,
      decimalsLimit,
      defaultValue,
      disabled = false,
      maxLength: userMaxLength,
      value: userValue,
      onValueChange,
      fixedDecimalLength,
      placeholder,
      decimalScale,
      prefix,
      suffix,
      intlConfig,
      step,
      min,
      max,
      disableGroupSeparators = false,
      disableAbbreviations = false,
      decimalSeparator: _decimalSeparator,
      groupSeparator: _groupSeparator,
      onChange,
      onFocus,
      onBlur,
      onKeyDown,
      onKeyUp,
      transformRawValue,
      ...props
    }: CurrencyInputProps,
    ref
  ) => {
    if (_decimalSeparator && isNumber(_decimalSeparator)) {
      throw new Error('decimalSeparator cannot be a number');
    }

    if (_groupSeparator && isNumber(_groupSeparator)) {
      throw new Error('groupSeparator cannot be a number');
    }

    const localeConfig = useMemo(() => getLocaleConfig(intlConfig), [intlConfig]);
    const decimalSeparator = _decimalSeparator || localeConfig.decimalSeparator || '';
    const groupSeparator = _groupSeparator || localeConfig.groupSeparator || '';

    if (
      decimalSeparator &&
      groupSeparator &&
      decimalSeparator === groupSeparator &&
      disableGroupSeparators === false
    ) {
      throw new Error('decimalSeparator cannot be the same as groupSeparator');
    }

    const formatValueOptions: Partial<FormatValueOptions> = {
      decimalSeparator,
      groupSeparator,
      disableGroupSeparators,
      intlConfig,
      prefix: prefix || localeConfig.prefix,
      suffix: suffix,
    };

    const cleanValueOptions: Partial<CleanValueOptions> = {
      decimalSeparator,
      groupSeparator,
      allowDecimals,
      decimalsLimit: decimalsLimit || fixedDecimalLength || 2,
      allowNegativeValue,
      disableAbbreviations,
      prefix: prefix || localeConfig.prefix,
      transformRawValue,
    };

    const formattedStateValue =
      defaultValue !== undefined && defaultValue !== null
        ? formatValue({ ...formatValueOptions, decimalScale, value: String(defaultValue) })
        : userValue !== undefined && userValue !== null
        ? formatValue({ ...formatValueOptions, decimalScale, value: String(userValue) })
        : '';

    const [stateValue, setStateValue] = useState(formattedStateValue);
    const [dirty, setDirty] = useState(false);
    const [cursor, setCursor] = useState(0);
    const [changeCount, setChangeCount] = useState(0);
    const [lastKeyStroke, setLastKeyStroke] = useState<string | null>(null);
    const inputRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(ref, () => inputRef.current as HTMLInputElement);

    /**
     * Process change in value
     */
    const processChange = (value: string, selectionStart?: number | null): void => {
      setDirty(true);

      const { modifiedValue, cursorPosition } = repositionCursor({
        selectionStart,
        value,
        lastKeyStroke,
        stateValue,
        groupSeparator,
      });

      const stringValue = cleanValue({ value: modifiedValue, ...cleanValueOptions });

      if (userMaxLength && stringValue.replace(/-/g, '').length > userMaxLength) {
        return;
      }

      if (stringValue === '' || stringValue === '-' || stringValue === decimalSeparator) {
        onValueChange && onValueChange(undefined, name, { float: null, formatted: '', value: '' });
        setStateValue(stringValue);
        // Always sets cursor after '-' or decimalSeparator input
        setCursor(1);
        return;
      }

      const stringValueWithoutSeparator = decimalSeparator
        ? stringValue.replace(decimalSeparator, '.')
        : stringValue;

      const numberValue = parseFloat(stringValueWithoutSeparator);

      const formattedValue = formatValue({
        value: stringValue,
        ...formatValueOptions,
      });

      if (cursorPosition !== undefined && cursorPosition !== null) {
        // Prevent cursor jumping
        let newCursor = cursorPosition + (formattedValue.length - value.length);
        newCursor = newCursor <= 0 ? (prefix ? prefix.length : 0) : newCursor;

        setCursor(newCursor);
        setChangeCount(changeCount + 1);
      }

      setStateValue(formattedValue);

      if (onValueChange) {
        const values: CurrencyInputOnChangeValues = {
          float: numberValue,
          formatted: formattedValue,
          value: stringValue,
        };
        onValueChange(stringValue, name, values);
      }
    };

    /**
     * Handle change event
     */
    const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
      const {
        target: { value, selectionStart },
      } = event;

      processChange(value, selectionStart);

      onChange && onChange(event);
    };

    /**
     * Handle focus event
     */
    const handleOnFocus = (event: React.FocusEvent<HTMLInputElement>): number => {
      onFocus && onFocus(event);
      return stateValue ? stateValue.length : 0;
    };

    /**
     * Handle blur event
     *
     * Format value by padding/trimming decimals if required by
     */
    const handleOnBlur = (event: React.FocusEvent<HTMLInputElement>): void => {
      const {
        target: { value },
      } = event;

      const valueOnly = cleanValue({ value, ...cleanValueOptions });

      if (valueOnly === '-' || valueOnly === decimalSeparator || !valueOnly) {
        setStateValue('');
        onBlur && onBlur(event);
        return;
      }

      const fixedDecimals = fixedDecimalValue(valueOnly, decimalSeparator, fixedDecimalLength);

      const newValue = padTrimValue(
        fixedDecimals,
        decimalSeparator,
        decimalScale !== undefined ? decimalScale : fixedDecimalLength
      );

      const numberValue = parseFloat(newValue.replace(decimalSeparator, '.'));

      const formattedValue = formatValue({
        ...formatValueOptions,
        value: newValue,
      });

      if (onValueChange) {
        onValueChange(newValue, name, {
          float: numberValue,
          formatted: formattedValue,
          value: newValue,
        });
      }

      setStateValue(formattedValue);

      onBlur && onBlur(event);
    };

    /**
     * Handle key down event
     *
     * Increase or decrease value by step
     */
    const handleOnKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
      const { key } = event;

      setLastKeyStroke(key);

      if (step && (key === 'ArrowUp' || key === 'ArrowDown')) {
        event.preventDefault();
        setCursor(stateValue.length);

        const currentValue =
          parseFloat(
            userValue !== undefined && userValue !== null
              ? String(userValue).replace(decimalSeparator, '.')
              : cleanValue({ value: stateValue, ...cleanValueOptions })
          ) || 0;
        const newValue = key === 'ArrowUp' ? currentValue + step : currentValue - step;

        if (min !== undefined && newValue < min) {
          return;
        }

        if (max !== undefined && newValue > max) {
          return;
        }

        const fixedLength = String(step).includes('.')
          ? Number(String(step).split('.')[1].length)
          : undefined;

        processChange(
          String(fixedLength ? newValue.toFixed(fixedLength) : newValue).replace(
            '.',
            decimalSeparator
          )
        );
      }

      onKeyDown && onKeyDown(event);
    };

    /**
     * Handle key up event
     *
     * Move cursor if there is a suffix to prevent user typing past suffix
     */
    const handleOnKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
      const {
        key,
        currentTarget: { selectionStart },
      } = event;
      if (key !== 'ArrowUp' && key !== 'ArrowDown' && stateValue !== '-') {
        const suffix = getSuffix(stateValue, { groupSeparator, decimalSeparator });

        if (suffix && selectionStart && selectionStart > stateValue.length - suffix.length) {
          /* istanbul ignore else */
          if (inputRef.current) {
            const newCursor = stateValue.length - suffix.length;
            inputRef.current.setSelectionRange(newCursor, newCursor);
          }
        }
      }

      onKeyUp && onKeyUp(event);
    };

    useEffect(() => {
      // prevent cursor jumping if editing value
      if (
        dirty &&
        stateValue !== '-' &&
        inputRef.current &&
        document.activeElement === inputRef.current
      ) {
        inputRef.current.setSelectionRange(cursor, cursor);
      }
    }, [stateValue, cursor, inputRef, dirty, changeCount]);

    /**
     * If user has only entered "-" or decimal separator,
     * keep the char to allow them to enter next value
     */
    const getRenderValue = () => {
      if (
        userValue !== undefined &&
        userValue !== null &&
        stateValue !== '-' &&
        (!decimalSeparator || stateValue !== decimalSeparator)
      ) {
        return formatValue({
          ...formatValueOptions,
          decimalScale: dirty ? undefined : decimalScale,
          value: String(userValue),
        });
      }

      return stateValue;
    };

    const inputProps: React.ComponentPropsWithRef<'input'> = {
      type: 'text',
      inputMode: 'decimal',
      id,
      name,
      className,
      onChange: handleOnChange,
      onBlur: handleOnBlur,
      onFocus: handleOnFocus,
      onKeyDown: handleOnKeyDown,
      onKeyUp: handleOnKeyUp,
      placeholder,
      disabled,
      value: getRenderValue(),
      ref: inputRef,
      ...props,
    };

    if (customInput) {
      const CustomInput = customInput;
      return <CustomInput {...inputProps} />;
    }

    return <input {...inputProps} />;
  }
);

CurrencyInput.displayName = 'CurrencyInput';

export default CurrencyInput;
