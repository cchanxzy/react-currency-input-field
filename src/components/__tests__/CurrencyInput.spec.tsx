import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CurrencyInput from '../CurrencyInput';

describe('<CurrencyInput/>', () => {
  const onValueChangeSpy = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Renders without error', () => {
    render(
      <CurrencyInput
        id="validationCustom01"
        name="inputName"
        placeholder="£1,000"
        className="form-control"
      />
    );
    const input = screen.getByRole('textbox');

    expect(input).toMatchSnapshot();

    expect(input).toHaveValue('');
  });

  it('Renders with default value', () => {
    render(<CurrencyInput defaultValue={1234.56} prefix="£" />);
    const input = screen.getByRole('textbox');

    expect(input).toMatchSnapshot();

    expect(input).toHaveValue('£1,234.56');
  });

  it('Renders with default value 0', () => {
    render(<CurrencyInput defaultValue={0} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('£0');
  });

  it('Renders with default value undefined', () => {
    render(<CurrencyInput defaultValue={undefined} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('Renders with default value null', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    render(<CurrencyInput defaultValue={null as any} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('Renders with default value 0 with decimalScale 2', () => {
    render(<CurrencyInput defaultValue={0} decimalScale={2} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('£0.00');
  });

  it('Renders with value prop', () => {
    render(<CurrencyInput value={49.99} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('£49.99');
  });

  it('Renders with value 0', () => {
    render(<CurrencyInput value={0} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('£0');
  });

  it('Renders with value undefined', () => {
    render(<CurrencyInput value={undefined} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('Renders with value null', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    render(<CurrencyInput value={null as any} prefix="£" />);

    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('Renders with value 0 with decimalScale 2', () => {
    render(<CurrencyInput value={0} prefix="£" decimalScale={2} />);

    expect(screen.getByRole('textbox')).toHaveValue('£0.00');
  });

  it('Renders with value 0.1 with decimalScale 2', () => {
    render(<CurrencyInput value={0.1} prefix="£" decimalScale={2} />);

    expect(screen.getByRole('textbox')).toHaveValue('£0.10');

    userEvent.type(screen.getByRole('textbox'), '{backspace}');

    expect(screen.getByRole('textbox')).toHaveValue('£0.1');
  });

  it('should go to end of string on focus', () => {
    render(<CurrencyInput defaultValue={123} />);
    userEvent.type(screen.getByRole('textbox'), '{arrowleft}4{arrowright}6');

    expect(screen.getByRole('textbox')).toHaveValue('12,436');
  });

  it('should allow value change with number', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '100');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('100', undefined, {
      float: 100,
      formatted: '£100',
      value: '100',
    });
  });

  it('should prefix 0 value', () => {
    render(<CurrencyInput prefix="£" value={0} onValueChange={onValueChangeSpy} />);

    expect(screen.getByRole('textbox')).toHaveValue('£0');
  });

  it('should allow 0 value on change', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '0');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('0', undefined, {
      float: 0,
      formatted: '£0',
      value: '0',
    });

    expect(screen.getByRole('textbox')).toHaveValue('£0');
  });

  it('should allow empty value', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} defaultValue={1} />);
    userEvent.clear(screen.getByRole('textbox'));

    expect(onValueChangeSpy).toHaveBeenLastCalledWith(undefined, undefined, {
      float: null,
      formatted: '',
      value: '',
    });

    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('should callback name as second parameter if name prop provided', () => {
    const name = 'inputName';
    render(<CurrencyInput name={name} prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '123');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('123', name, {
      float: 123,
      formatted: '£123',
      value: '123',
    });
  });

  it('should not allow invalid characters', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), 'hello');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith(undefined, undefined, {
      float: null,
      formatted: '',
      value: '',
    });

    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('should ignore invalid characters', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '£123hello');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('123', undefined, {
      float: 123,
      formatted: '£123',
      value: '123',
    });

    expect(screen.getByRole('textbox')).toHaveValue('£123');
  });

  it('should clear decimal point only input', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '.');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith(undefined, undefined, {
      float: null,
      formatted: '',
      value: '',
    });

    fireEvent.focusOut(screen.getByRole('textbox'));
    expect(screen.getByRole('textbox')).toHaveValue('');
  });

  it('should allow .3 decimal inputs', () => {
    render(<CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '.3');

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('.3', undefined, {
      float: 0.3,
      formatted: '£0.3',
      value: '.3',
    });

    fireEvent.focusOut(screen.getByRole('textbox'));
    expect(screen.getByRole('textbox')).toHaveValue('£0.3');
  });

  it('should call onChange', () => {
    const onChangeSpy = jest.fn();
    render(<CurrencyInput prefix="£" onChange={onChangeSpy} />);
    userEvent.type(screen.getByRole('textbox'), '123');

    expect(onChangeSpy).toBeCalledTimes(3);

    expect(screen.getByRole('textbox')).toHaveValue('£123');
  });

  it('should call onBlur', () => {
    const onBlurSpy = jest.fn();
    render(<CurrencyInput prefix="£" onBlur={onBlurSpy} />);
    userEvent.click(screen.getByRole('textbox'));
    fireEvent.focusOut(screen.getByRole('textbox'));

    expect(onBlurSpy).toBeCalledTimes(1);
  });

  it('should call onFocus', () => {
    const onFocusSpy = jest.fn();
    render(<CurrencyInput onFocus={onFocusSpy} />);
    fireEvent.focusIn(screen.getByRole('textbox'));

    expect(onFocusSpy).toBeCalledTimes(1);
  });

  it('should call onKeyDown', () => {
    const onKeyDownSpy = jest.fn();
    render(<CurrencyInput onKeyDown={onKeyDownSpy} />);
    userEvent.type(screen.getByRole('textbox'), '1');

    expect(onKeyDownSpy).toBeCalledTimes(1);
  });

  it('should call onKeyUp', () => {
    const onKeyUpSpy = jest.fn();
    render(<CurrencyInput onKeyUp={onKeyUpSpy} />);
    userEvent.type(screen.getByRole('textbox'), '1');

    expect(onKeyUpSpy).toBeCalledTimes(1);
  });
});
