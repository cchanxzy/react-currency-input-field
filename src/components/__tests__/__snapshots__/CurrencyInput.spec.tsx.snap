// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyInput/> Renders with default value 1`] = `
<input
  inputmode="decimal"
  type="text"
  value="£1,234.56"
/>
`;

exports[`<CurrencyInput/> Renders without error 1`] = `
<input
  class="form-control"
  id="validationCustom01"
  inputmode="decimal"
  name="inputName"
  placeholder="£1,000"
  type="text"
  value=""
/>
`;
